PolyEngine/Core/Src/Queue.hpp

```diff
@@ -9,17 +9,88 @@ namespace Poly
 	{
 	public:
 		Queue() = default;
-		Queue(size_t size) : Data(size) {};
+		Queue(size_t size) : Data(size), RingCap(size), BuffCap(size)
+		{
+			for (size_t i = 0; i < size; i++)
+			{
+				Data.PushBack(T());
+			}
+		}
 
-		bool Empty() const { return Data.IsEmpty(); }
-		size_t Size() const { return Data.GetSize(); }
-		const T& Front() const { return Data.GetData()[0]; }
-		const T& Back() const { return Data.GetData()[Data.GetSize() - 1]; };
-		T& Front() { return Data.GetData()[0]; }
-		T& Back() { return Data.GetData()[Data.GetSize() - 1]; };
-		void Push(const T &obj) { Data.PushBack(obj); }
-		void Pop() { Data.PopFront(); }
+		bool Empty() const { return BuffSize == 0; }
+		size_t Size() const { return BuffSize; }
+		const T& Front() const { return Data.GetData()[Head]; }
+		const T& Back() const 
+		{
+			if (RingCap == BuffCap)
```

Comment:
You can move this to private method `T& BackImpl() const` and then call this method in both versions of Back